### PR TITLE
try add osx_arm64 build

### PR DESCRIPTION
**Pull request by SrNetoChan**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

This is necessary for building QGIS in osx_arm64 and linux_aarch64
https://github.com/conda-forge/qgis-feedstock/pull/289

### CODE REVIEW

**Comment by SrNetoChan** [comment]
@conda-forge-admin, please rerender

**Comment by SrNetoChan** [comment]
any ideas @gillins ?

**Comment by SrNetoChan** [comment]
It give and status 126 Error. Googling around it seems related to some execution permissions on the file 
Users/runner/miniforge3/conda-bld/qscintilla2_1672709146979/work/conda_build.sh

**Comment by gillins** [comment]
Thanks @SrNetoChan I think it might be trying to run Python but for the wrong CPU type when cross compiling. I think you might need these lines in the `build` section:
```
    - python                                # [build_platform != target_platform]
    - cross-python_{{ target_platform }}    # [build_platform != target_platform]
```
See https://conda-forge.org/blog/posts/2020-10-29-macos-arm64/#how-to-add-a-osx-arm64-build-to-a-feedstock and the changes I needed to get TuiView working: https://github.com/conda-forge/tuiview-feedstock/pull/51/files#diff-f3725a55bf339595bf865fec73bda8ac99f283b0810c205442021f29c06eea9a

**Comment by SrNetoChan** [comment]
It's complaining about sip build both in osx_arm and aarch64

**Comment by gillins** [comment]
Can you bring in the ``if [[ "${CONDA_BUILD_CROSS_COMPILATION:-}" == "1" ]]; then`` parts from https://github.com/conda-forge/pyqtwebkit-feedstock/blob/main/recipe/build.sh ? Hopefully that should fix it.

**Comment by SrNetoChan** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml
Comment (gillins) [inline recipe/meta.yaml:95]: Also need:
```
- sip                                    # [build_platform != target_platform]
```

**Comment by gillins** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml
Comment (gillins) [reply recipe/meta.yaml:95]: `pyqtwebkit` also needs this:
```
- pyqt-builder                           # [build_platform != target_platform]
```
Might be worth a try?

Comment (SrNetoChan) [reply recipe/meta.yaml:95]: With that if platform != ... Right?

Comment (gillins) [reply recipe/meta.yaml:95]: Yep - that's right only for cross compilation. No idea what all this stuff does, but hopefully if we copy enough of the `pyqtwebkit` recipe we'll get it working....

Comment (SrNetoChan) [reply recipe/meta.yaml:95]: Last one was different error, so you are getting there. I have no idea of what I am doing :-p

**Comment by gillins** [comment]
OK that didn't make much difference. I wonder if you could try syncing up this part of the `build.sh`: https://github.com/conda-forge/pyqtwebkit-feedstock/blob/main/recipe/build.sh#L9-L34 ? I'm not sure if it will help but at least we will be similar to something that works...

**Comment by SrNetoChan** [comment]
No change. I will revert that too. Meanwhile, I found this:

https://www.mail-archive.com/qscintilla@riverbankcomputing.com/msg01077.html

Exatly the same error.

File: recipe/build.sh
Comment (gillins) [inline recipe/build.sh:70]: maybe add `--qt-shared`? `pyqtwebkit` has this... If that doesn't work, I'm afraid I'm out of ideas,

**Comment by gillins** [comment]
@hmaarrfk do you have any thoughts on this one? We are kind of clutching at straws here...

**Comment by SrNetoChan** [comment]
The error both in arm or aarch is  happens during Building Python bindings

The error is:
```
build.py: /home/conda/feedstock_root/build_artifacts/qscintilla2_1673030184046/work/Python/sip/qscimodcommon.sip: line 42: column 9: 'QtCore/QtCoremod.sip' could not be found
/home/conda/feedstock_root/build_artifacts/qscintilla2_1673030184046/work/Python/sip/qscimodcommon.sip: line 43: column 9: 'QtGui/QtGuimod.sip' could not be found
/home/conda/feedstock_root/build_artifacts/qscintilla2_1673030184046/work/Python/sip/qscimodcommon.sip: line 44: column 9: 'QtWidgets/QtWidgetsmod.sip' could not be found
/home/conda/feedstock_root/build_artifacts/qscintilla2_1673030184046/work/Python/sip/qscimodcommon.sip: line 46: column 6: 'PyQt_Printer' is not a known qualifier
/home/conda/feedstock_root/build_artifacts/qscintilla2_1673030184046/work/Python/sip/qsciprinter.sip: line 21: column 6: 'PyQt_Printer' is not a known qualifier

```

Googling for the error I found some references [1][2] to exactly the same error while building Qscintilla2, but no solution, only the hint that this kind of error happens when the builder is using the wrong python/sip within the system. In other cases it's suggested to install also the pyqt5-dev tools 

https://stackoverflow.com/questions/18922009/pyqt-unable-to-find-qtcoremod-sip
https://www.mail-archive.com/qscintilla@riverbankcomputing.com/msg01077.html
https://github.com/qgis/QGIS/issues/19466

@conda-forge/arm-arch team any ideas?

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

basically cross compilation is just hard with qt.... so I avoid it as much as I can.

**Comment by hmaarrfk** [comment]
support is very limited for cross compilation in 5.X from qt.

**Comment by gillins** [comment]
Thanks @hmaarrfk, very interesting. So I guess osx_arm64 is likely to be out of reach for now...
Our original goal was to build qgis which takes about 4.5 hours ATM so we were hoping cross compiling would be possible. I think it also uses sip.

**Comment by SrNetoChan** [comment]
I don't have an Mac M1 but maybe we could find someone with one to try building it locally?

**Comment by SrNetoChan** [comment]
What would I need to run? Something like this?
```
conda mambabuild ./recipe -m ./.ci_support/osx_arm64_python3.9.____cpython.yaml
```

**Comment by SrNetoChan** [comment]
@gillins I got access to a MacOs M1 machine and built this feedstock there. The files are here:

https://anaconda.org/afneto/qscintilla2/files

If we get someone to upload these and merge the https://github.com/conda-forge/qjson-feedstock/pull/15 we can try to build QGIS again.

**Comment by hmaarrfk** [comment]
cross compiling works a little better for osx-arch than it does for linux aarch. I wouldn't give up just yet. maybe @jschueller  has better insights.

**Comment by SrNetoChan** [comment]
It's interesting that the error was exactly the same in aarch and arm.

Also, it was nice to see that the recipe worked fine building locally on a mac m1.

We will wait for your insites

File: recipe/build.sh
Comment (hmaarrfk) [inline recipe/build.sh:77]: can yo utry commenting this out?

**Comment by gillins** [comment]
Yep would be good to try a few more things and see if we can get it going. Building locally each time there is an update required is going to be a pain.... esp. for qgis.

File: recipe/build.sh
Comment (SrNetoChan) [reply recipe/build.sh:77]: Done, let's see how it goes.

**Comment by hmaarrfk** [comment]
@pkgw do you think they might be hitting the same bug you addressed for cross python:
```
2023-01-11T16:46:40.6713780Z + sip-build --no-make --qsci-features-dir ../src/features --qsci-include-dir ../src --qsci-library-dir ../src --api-dir /Users/runner/miniforge3/conda-bld/qscintilla2_1673454887107/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/qsci/api/python
2023-01-11T16:46:40.6883700Z /Users/runner/miniforge3/conda-bld/qscintilla2_1673454887107/_build_env/bin/sip-build: line 4: import: command not found
2023-01-11T16:46:40.6898830Z /Users/runner/miniforge3/conda-bld/qscintilla2_1673454887107/_build_env/bin/sip-build: line 5: import: command not found
2023-01-11T16:46:40.7054270Z from: can't read /var/mail/sipbuild.tools.build
2023-01-11T16:46:41.4955060Z /Users/runner/miniforge3/conda-bld/qscintilla2_1673454887107/_build_env/bin/sip-build: line 10: syntax error near unexpected token `('
2023-01-11T16:46:41.4956860Z /Users/runner/miniforge3/conda-bld/qscintilla2_1673454887107/_build_env/bin/sip-build: line 10: `    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])'
```

**Comment by hmaarrfk** [comment]
FYI, it might be related to: https://github.com/conda-forge/cross-python-feedstock/pull/65

**Comment by SrNetoChan** [comment]
@hmaarrfk should I revert the last commit? It seems that the current error is fixed by it, although we will hit the next one.

**Comment by hmaarrfk** [comment]
I think you can revert, and maybe add a reference to the issue I linked to. it shouldn't be necessary, but you did well in forcing the correct flags.

**Comment by hmaarrfk** [comment]
I'm really not sure what the best way forward for you all is.

1. Invoking cfep03 https://github.com/conda-forge/cfep/blob/main/cfep-03.md isn't the "easiest" thing. It also isn't meant for cases like this where we are simply having trouble cross compiling.
2. Wait to get help from somebody more knowledgable. 

I'm sorry i can't help anymore.

**Comment by SrNetoChan** [comment]
Interesting that the problem was exactly the same both in aarch64 and arm64. I would say it's needed some patch on sip builder configure.py.

**Comment by SrNetoChan** [comment]
Since the error in cross compiling for OSX arm64 was the same as when trying cross compilation on linux, I was trying to build aarch64 locally on my kubuntu 22.04 machine. But I was stopped even before that error appears.  I am getting this:

```
++ qmake -v
++ sed -n 's/.*Qt version \([0-9])*\).*/\1/p'
/home/aneto/mambaforge/envs/debug_qscintilla2/conda-bld/qscintilla2_1673738274569/work/conda_build.sh: line 14: /home/aneto/mambaforge/envs/debug_qscintilla2/conda-bld/qscintilla2_1673738274569/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/bin/qmake: cannot execute binary file: Exec format error
+ QT_MAJOR_VER=
Traceback (most recent call last):
  File "/home/aneto/mambaforge/envs/debug_qscintilla2/bin/conda-mambabuild", line 10, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/aneto/mambaforge/envs/debug_qscintilla2/lib/python3.11/site-packages/boa/cli/mambabuild.py", line 256, in main
    call_conda_build(action, config)
  File "/home/aneto/mambaforge/envs/debug_qscintilla2/lib/python3.11/site-packages/boa/cli/mambabuild.py", line 228, in call_conda_build
    result = api.build(
             ^^^^^^^^^^
  File "/home/aneto/mambaforge/envs/debug_qscintilla2/lib/python3.11/site-packages/conda_build/api.py", line 180, in build
    return build_tree(
           ^^^^^^^^^^^
  File "/home/aneto/mambaforge/envs/debug_qscintilla2/lib/python3.11/site-packages/conda_build/build.py", line 3097, in build_tree
    packages_from_this = build(metadata, stats,
                         ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/aneto/mambaforge/envs/debug_qscintilla2/lib/python3.11/site-packages/conda_build/build.py", line 2209, in build
    utils.check_call_env(cmd, env=env, rewrite_stdout_env=rewrite_env,
  File "/home/aneto/mambaforge/envs/debug_qscintilla2/lib/python3.11/site-packages/conda_build/utils.py", line 402, in check_call_env
    return _func_defaulting_env_to_os_environ('call', *popenargs, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/aneto/mambaforge/envs/debug_qscintilla2/lib/python3.11/site-packages/conda_build/utils.py", line 382, in _func_defaulting_env_to_os_environ
    raise subprocess.CalledProcessError(proc.returncode, _args)
subprocess.CalledProcessError: Command '['/bin/bash', '-o', 'errexit', '/home/aneto/mambaforge/envs/debug_qscintilla2/conda-bld/qscintilla2_1673738274569/work/conda_build.sh']' returned non-zero exit status 126.
```

Seems that bash is not being able to run qmake because it does not consider qmake to have execution permissions.

I am running the following command inside a pre created-environment:

    conda mambabuild ./recipe -m ./.ci_support/linux_aarch64_python3.8.____cpython.yaml

**Comment by gillins** [comment]
I think this is because we are building aarch64 natively now so it is probably trying to run arm binaries on your x86 machine... I wonder if you set aarch64 back to azure and re-render it might work better. Also I think you can just run the feedstock script `build_locally.py`. @hmaarrfk will know more.

File: conda-forge.yml
Comment (hmaarrfk) [inline conda-forge.yml]:
```suggestion
  linux_aarch64: linux_64
```

You are now building linux aarch64 "natively", well really, using "emulation + qemu".

You can go back to cross compilation with this change + rerender

Comment (hmaarrfk) [reply conda-forge.yml]: you can also build locally with


```
python build-locally.py
```

or:

```
conda debug -m .ci_support/THECONFIG.yaml recipe
```

**Comment by SrNetoChan** [comment]
Ah! Always learning. Makes sense. I will try it tomorrow (monday).

**Comment by SrNetoChan** [comment]
@gillins @hmaarrfk still the same error or similar after changing to cross compiling and rerendering. Seems I am failing to create the exact same scenario as in azure. :-(

**Comment by hmaarrfk** [comment]
>  still the same error or similar after changing to cross compiling and rerendering. Seems I am failing to create the exact same scenario as in azure. :-(

i don't understand. what are you trying to compile? Linux, OSX? can you provide the exact commands you are running.

**Comment by SrNetoChan** [comment]
> > still the same error or similar after changing to cross compiling and rerendering. Seems I am failing to create the exact same scenario as in azure. :-(
> 
> i don't understand. what are you trying to compile? Linux, OSX? can you provide the exact commands you are running.

I run a linux machine. Since the error currently in osx arm64 was exactly the same in linux aarch64 before we buildi it natively. I was trying to going back to cross'compiling on my local machine to try figure out waht may be hapening on the OSX front. 

I also have access to a osx arm computer, where I was able to build osx arm64 without problems, but I think it ended up building natevely, so I donºt know if I can replicate the environment of azure osx

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

**Comment by hmaarrfk** [comment]
Can you try now?

If you can generate the log

```
python build-locally.py 2>&1 | tee log.txt
```

that would be helpful

**Comment by hmaarrfk** [comment]
you can also try:
```
conda debug -m .ci_support/THE_BUILD_YOUWANT recipe
```

**Comment by SrNetoChan** [comment]
This was what I got with 

python build-locally.py 2>&1 | tee log.txt

[log.txt](https://github.com/conda-forge/qscintilla2-feedstock/files/10448232/log.txt)

**Comment by hmaarrfk** [comment]
On linux, because we can, we typically build with "emulation" which means that the binaries in the `PREFIX` environment are also available to the build system.

It is likely trying to run the aarch executable.

On ubuntu, I think you need to install `qemu-user-static`. It might be a similar package on other platforms.

**Comment by SrNetoChan** [comment]
So I should install that on my machine? Using apt for example?

**Comment by hmaarrfk** [comment]
yes, something like
```
sudo apt install qemu-user-static
```

**Comment by hmaarrfk** [comment]
basically, docker looks for those files, then uses those files to emulate "arch" when it detects binaries with that format.

**Comment by SrNetoChan** [comment]
> sudo apt install qemu-user-static

@hmaarrfk great that did the difference, now I am in the same situation as azure.. Let's see if I can leverage that to figure out what might be wrong.

**Comment by SrNetoChan** [comment]
So what I have found so far. I run the debug session and started the build_env that takes me to this folder:

```
/home/aneto/mambaforge/conda-bld/debug_1674134804089/work
```
with the following active environemnt:
```
/home/aneto/mambaforge/conda-bld/debug_1674134804089/_build_env
```
I tried to run the conda_build.sh file and got the exact same error as before. Something like:

```
column 9: 'QtCore/QtCoremod.sip' could not be found
column 9: 'QtGui/QtGuimod.sip' could not be found
column 9:  'QtWidgets/QtWidgetsmod.sip' could not be found
column 6: 'PyQt_Printer' is not a known qualifier
column 6: 'PyQt_Printer' is not a known qualifier
```

These are all import in the following file:

```
/home/aneto/mambaforge/conda-bld/debug_1674134804089/work/Python/sip/qscimodcommon.sip
```

I have tried to find there QtCore/QtCoremod.sip files and the are all in this folder:

```
/home/aneto//mambaforge/conda-bld/debug_1674134804089/_build_env/lib/python3.10/site-packages/PyQt5/bindings/QtCore
```

Nevertheless, when running conda_build.sh the folder set for SIP_DIR is that one, but 

```
SIP_DIR=/home/aneto/mambaforge/conda-bld/debug_1674134804089/_h_env/lib/python3.10/site-packages/PyQt5/bindings
```
which does not exist... actually the _h_env/lib/python3.10/site-packages/ folder is completely empty

notice the _h_env instead of _build_env

Are these the host environment and the build environment? So, I wonder if we need to install something extra in the host if the SIP_DIR path is wrong.

**Comment by hmaarrfk** [comment]
Maybe @conda-forge/pyqt can help

**Comment by SrNetoChan** [comment]
@conda-forge-admin, please rerender

**Comment by SrNetoChan** [comment]
Not pretty, but it workd locally for linux, let's see if it does the trick for OSX arm too.

**Comment by SrNetoChan** [comment]
Hurrey!!!!

@hmaarrfk and @gillins thanks for all your help. I have learned quite a lot.

See if the build.sh change could be made in another way, the repeated code looks ugly.

**Comment by hmaarrfk** [comment]
Great work! Sorry I didn't want to invoke CFEP03. I think this way you are more empowered to move at your speed in the future!

**Comment by SrNetoChan** [comment]
> Great work! Sorry I didn't want to invoke CFEP03. I think this way you are more empowered to move at your speed in the future!

Definitely better this way, as it fixes it for good. I wonder if native arm compiling will be available in the future.

**Comment by hmaarrfk** [comment]
Why not cross compile linux-aarch64 too? it is "faster" and can be a good quality of life improvement for future versions.

**Comment by SrNetoChan** [comment]
Since you said it was complicated and it worked natively... No strong feelings about it.

**Comment by hmaarrfk** [comment]
oh but you fixed things! you went through the complications!

**Comment by SrNetoChan** [comment]
In that case, let's move it back to cross compiling also for linux, so it's much faster.

**Comment by gillins** [comment]
Thanks @SrNetoChan ! That was an inspired bit work. Is this ready to merge yet, or do you need to make the Linux cross compiling change?

**Comment by SrNetoChan** [comment]
Well lets follow @hmaarrfk advice and make it faster to compile

**Comment by SrNetoChan** [comment]
@conda-forge-admin, please rerender

**Comment by SrNetoChan** [comment]
@gillins the change is done, you can add that tag that allows to merge if all checks are good. it shouldn't take long tho.

**Comment by gillins** [comment]
I'm not a maintainer on this feedstock so will have to wait for @conda-forge/qscintilla2 to merge for us - I think it's all good to go now - thanks!

**Comment by hmaarrfk** [comment]
given your intimate knowledge of the package ,you might want to ask to be a maintainer too.

**Comment by SrNetoChan** [comment]
@ocefpaf could you please merge this one?

**Comment by SrNetoChan** [comment]
@conda-forge-admin, please ping team

**Comment by gillins** [comment]
@SrNetoChan do you want to add yourself as a maintainer as part of this PR?

**Comment by SrNetoChan** [comment]
Not sure how active I will be able to be, and... I really don't know what I am doing :-)

**Comment by SrNetoChan** [comment]
I have add myself as maintainer.

**Comment by matthiasdiener** [comment]
@conda-forge-admin, please rerender

**Comment by SrNetoChan** [comment]
@matthiasdiener I forgot to rerender. thanks!

**Review by matthiasdiener — APPROVED** [review]
Thank you for your work on this @SrNetoChan !

**Comment by SrNetoChan** [comment]
@matthiasdiener can you please merge it. I don't have permissions for that.

**Pull request merged**
